Kernel/Arch/riscv64/PCI/Initializer.cpp

```diff
@@ -4,17 +4,159 @@
+#include <Kernel/Memory/MemoryManager.h>
+#include <Userland/Libraries/LibDeviceTree/FlattenedDeviceTree.h>
 
 namespace Kernel::PCI {
 
 bool g_pci_access_io_probe_failed { false };
-bool g_pci_access_is_disabled_from_commandline { true };
+bool g_pci_access_is_disabled_from_commandline;
 
 void initialize()
 {
-    TODO_RISCV64();
+    g_pci_access_is_disabled_from_commandline = kernel_command_line().is_pci_disabled();
+    if (g_pci_access_is_disabled_from_commandline)
+        return;
+
+    new Access();
+
+    // https://github.com/devicetree-org/devicetree-specification/releases/download/v0.4/devicetree-specification-v0.4.pdf
+    // https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/pci/pci-bus-common.yaml
+    // https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/pci/pci-host-bridge.yaml
+
+    // The pci controllers are usually in /soc/pcie?@XXXXXXXX
+    // FIXME: They can also appear in the root node, or any simple-bus other than soc
+    auto const& device_tree = DeviceTree::get();
+
+    auto maybe_soc = device_tree.get_child("soc"sv);
+    if (!maybe_soc.has_value()) {
+        dmesgln("PCI: No `soc` node found in the device tree, PCI initialization will be skipped");
+        return;
+    }
+
+    auto const& soc = maybe_soc.value();
+
+    enum class ControllerCompatible {
+        CAM,
+        ECAM,
+    };
+
+    // These properties must be present
+    auto soc_address_cells = soc.get_property("#address-cells"sv).value().as<u32>();
+    [[maybe_unused]] auto soc_size_cells = soc.get_property("#size-cells"sv).value().as<u32>();
+
+    Optional<u32> domain_counter;
+    for (auto const& entry : soc.children()) {
+        if (!entry.key.starts_with("pci"sv))
+            continue;
+        auto const& node = entry.value;
+
+        if (auto device_type = node.get_property("device_type"sv); !device_type.has_value() || device_type.value().as_string() != "pci"sv) {
+            dmesgln("PCI: PCI named device tree entry {} not a PCI controller, got device type {} instead", entry.key, device_type.has_value() ? device_type.value().as_string() : "<None>"sv);
+            continue;
+        }
+
+        auto maybe_compatible = node.get_property("compatible"sv);
+        if (!maybe_compatible.has_value()) {
+            dmesgln("PCI: PCI controller {} did not provide a compatible string, rejecting", entry.key);
+            continue;
+        }
+        auto compatible = maybe_compatible.value().as_strings();
+        ControllerCompatible controller_compatibility;
+        if (compatible.contains_slow("pci-host-ecam-generic"sv)) {
+            controller_compatibility = ControllerCompatible::ECAM;
+        } else if (compatible.contains_slow("pci-host-cam-generic"sv)) {
+            controller_compatibility = ControllerCompatible::CAM;
+            // FIXME: Implement CAM
+            dmesgln("PCI: PCI controller {} is CAM, which is not supported", entry.key);
```

Comment:
Not sure if we should have a specific case for `pci-host-cam-generic`.
Most real hardware doesn't use the `pci-host-[e]cam-generic` compatibles.